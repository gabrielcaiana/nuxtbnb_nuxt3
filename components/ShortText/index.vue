<script setup>
const props = defineProps({
  text: {
    type: String,
    required: true,
  },
  target: {
    type: Number,
    required: true,
  },
});

const isExpanded = ref(false);
const chunks = ref([]);

const isTooLong = computed(() => chunks.value.length === 2);
const displayText = computed(() => {
  if (!isTooLong.value || isExpanded.value) return chunks.value.join(" ");
  return `${chunks.value[0]}...`;
});

const getChunks = () => {
  const position = props.text.indexOf(" ", props.target);
  if (props.text.length <= props.target || position === -1) return [props.text];
  return [props.text.substring(0, position), props.text.substring(position)];
};

chunks.value = getChunks();
</script>

<template>
  <span>
    {{ displayText }}
    <button
      v-if="isTooLong && !isExpanded"
      @click="isExpanded = true"
      class="link"
      type="button"
    >
      read more
    </button>
    <button
      v-if="isTooLong && isExpanded"
      @click="isExpanded = false"
      class="link"
      type="button"
    >
      read less
    </button>
  </span>
</template>

<style scoped>
.link {
  color: blue;
  background-color: #fff;
  border: none;
  text-decoration: underline;
  cursor: pointer;
}
.link:focus {
  border: none;
  outline: none;
}
</style>
