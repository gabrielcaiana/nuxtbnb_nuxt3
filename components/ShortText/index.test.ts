import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import ShortText from "./index.vue";
import text from "~/test/fixtures/reviews";

describe("ShortText.vue", () => {
  const wrapper = mount(ShortText, {
    props: {
      text: text.comment,
      target: 150,
    },
  });

  const component = wrapper.find('[datatest-id="short-text"]');

  it("should render the correct component", () => {
    expect(component.exists()).toBeTruthy();
    expect(component).toMatchSnapshot();
  });

  it("should display full text and less text", async () => {
    await wrapper.find('[datatest-id="read-more"]').trigger("click");
    expect(component.text()).toBe(
      `Cute house in a nice neighborhood. Close to downtown and a few blocks from Sanders Beach which was really nice. A few minor notes—the neighbor is an early  riser and started mowing the lawn right next to our bedroom windows at 7:30am. I also thought it was weird that we had to empty the dishwasher and take out all of the garbage before we checked out at 10am. I normally load and run the dishwasher before I leave but was unable to do so since the dishwasher had to be emptied and everything in it’s originally place. It’s minor but since the cleaning fee is more then I paid my maid for deep cleaning I would have thought I could load the dishwasher and run it and wouldn’t have to take the garbage out. Otherwise cute house, good neighborhood, clean, and comfy.  read less`
    );

    await wrapper.find('[datatest-id="read-less"]').trigger("click");

    expect(component.text()).toBe(
      "Cute house in a nice neighborhood. Close to downtown and a few blocks from Sanders Beach which was really nice. A few minor notes—the neighbor is an early...  read more"
    );
  });
});
