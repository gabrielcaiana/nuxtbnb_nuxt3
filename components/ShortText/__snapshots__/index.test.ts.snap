// Vitest Snapshot v1

exports[`ShortText.vue > should render the correct component 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <span
    data-v-581524dc=""
    datatest-id="short-text"
  >
    Cute house in a nice neighborhood. Close to downtown and a few blocks from Sanders Beach which was really nice. A few minor notesâ€”the neighbor is an early... 
    <button
      class="link"
      data-v-581524dc=""
      datatest-id="read-more"
      type="button"
    >
       read more 
    </button>
    <!--v-if-->
  </span>,
}
`;
